modular_skyrat/modules/ashwalkers/code/buildings/ash_farming.dm

```diff
@@ -1,141 +1,194 @@
-	if(istype(attacking_object, /obj/item/grown))
-		var/obj/item/grown/attacking_grown = attacking_object
-		if(attacking_grown.seed)
-			try_seeding(attacking_grown.seed, attacking_grown, user)
-			return
-	else if(istype(attacking_object, /obj/item/food/grown))
-		var/obj/item/food/grown/attacking_food = attacking_object
-		if(attacking_food.seed)
-			try_seeding(attacking_food.seed, attacking_food, user)
-			return
+/datum/component/simple_farm
+	///whether you can actually farm it at the moment
+	var/allow_plant = FALSE
+	///whether we limit the amount of plants you can have per turf
+	var/one_per_turf = TRUE
+	///the reference to the movable parent the component is attached to
+	var/atom/atom_parent
+	///the amount of pixels shifted (x,y)
+	var/list/pixel_shift = 0
+
+/datum/component/simple_farm/Initialize(set_plant = FALSE, set_turf_limit = TRUE, list/set_shift = list(0, 0))
+	//we really need to check if its movable
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	atom_parent = parent
+	//important to allow people to just straight up set allowing to plant
+	allow_plant = set_plant
+	one_per_turf = set_turf_limit
+	pixel_shift = set_shift
+	//now lets register the signals
+	RegisterSignal(atom_parent, COMSIG_PARENT_ATTACKBY, .proc/check_attack)
+	RegisterSignal(atom_parent, COMSIG_PARENT_EXAMINE, .proc/check_examine)
+
+/datum/component/simple_farm/Destroy(force, silent)
+	//lets not hard del
+	UnregisterSignal(atom_parent, list(COMSIG_PARENT_ATTACKBY, COMSIG_PARENT_EXAMINE))
+	atom_parent = null
 	return ..()
 
-/turf/open/misc/asteroid/basalt/lava_land_surface/proc/try_seeding(obj/item/seeds/planting_seed, obj/targeted_delete, mob/user)
-	var/obj/structure/flora/ash_farming/find_farm = locate() in contents
-	if(find_farm)
-		to_chat(user, span_warning("There can only be one farm in a hole at a time!"))
-		return
-	var/planting_chance = HAS_TRAIT(user, TRAIT_PRIMITIVE) ? 100 : 40
-	if(!dug)
-		to_chat(user, span_warning("You cannot grow plants on [src] without it being dug first!"))
-		return
-	if(!planting_seed)
-		to_chat(user, span_warning("[targeted_delete] does not have a seed, it cannot be grown!"))
-		return
-	to_chat(user, span_notice("You begin planting..."))
-	if(!do_after(user, 5 SECONDS, target = src))
-		to_chat(user, span_warning("You interrupt your planting!"))
-		return
-	if(!prob(planting_chance))
-		to_chat(user, span_warning("[targeted_delete] breaks in your hands!"))
-		qdel(targeted_delete)
-		return
-	var/obj/structure/flora/ash_farming/new_farm = new /obj/structure/flora/ash_farming(src)
-	new_farm.planted_seed = planting_seed
-	new_farm.planted_type = planting_seed.type
-	new_farm.appearance_naming()
-	qdel(targeted_delete)
-
-/obj/structure/flora/ash_farming
-	name = "ash flora"
-	desc = "A plant that has mutated to adapt to the environment."
-	///the seed that is "within" the structure, vars are pulled from it
+/datum/component/simple_farm/proc/check_attack(datum/source, obj/item/attacking_item, mob/user)
```

Comment:
Documentation please (saying what the arguments are too, mostly for stuff like `datum/source`, but it's just a good practice in general.